internal/operators/operators.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operators
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	columnhelpers "github.com/inspektor-gadget/inspektor-gadget/internal/column-helpers"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type Runner interface {
+	ID() string
+	Columns() columnhelpers.Columns
+	Gadget() gadgets.Gadget
+	Context() context.Context
+	Operators() Operators
+	Logger() logger.Logger
+}
+
+type Operator interface {
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// Params will return global params (required) for this operator
+	Params() params.Params
+
+	// PerGadgetParams will return params (required) per gadget instance of the operator
+	PerGadgetParams() params.Params
+
+	// Dependencies can list other operators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.Gadget) bool
+
+	// Init allows the operator to initialize itself
+	Init(params params.Params) error
+
+	// Cleanup allows the operator to clean up stuff prior to exiting
+	Cleanup() error
+
+	// PreGadgetRun is called before a gadget is started; the operator must return something that implements operator
+	// This is useful to create a context for an operator by wrapping it.
+	// Params given here are the ones returned by PerGadgetParams()
```

Comment:
I find this confusing. Why this operation has to return something that implements this interface? Is this possible to use a different interface for it? I'm wondering what's the purpose of this so see what other alternatives we have.

Reply:
Hehe, I already talked with Jose about this and agree that it's confusing. It simplifies how operators are stored a bit at the cost of being that bit confusing (they are stored in an array of this interface and replaced for the specific gadget run once PreGadgetRun is called). Having the same interface allows to still access Name(), Description(), Params() and so on - and some operators might not need a wrapper at all, if they don't need to store per-run data. I'll think more about it.

Reply:
I understand the optimisation, but AFAIU, `Name()`, `Description()` and `Params()` are not used by the object returned by `PreGadgetRun`. Therefore, we are not actually taking advantage of such optimisation. As we already discussed, I would separate this into two interfaces and keep things simple and easy to understand. Later, when everybody is used to the code, we can start introducing optimisations.

Reply:
I changed this now to have a new OperatorInstance interface that is used as return type. But instead of having `PreGadgetRun()` instantiate, there's now a new `Instanciate()` function. This is required for handling events asynchronously later on.

Reply:
I propose to resolve this discussion after having implemented what we agreed [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/1281#discussion_r1090737641).